internal/declarative/v2/reconciler.go

```diff
@@ -403,15 +404,23 @@ func (r *Reconciler) doPreDelete(ctx context.Context, clnt Client, obj Object) e
 
 func (r *Reconciler) renderTargetResources(
 	ctx context.Context,
+	clnt client.Client,
 	converter ResourceToInfoConverter,
 	obj Object,
 	spec *Spec,
 ) ([]*resource.Info, error) {
-	if !obj.GetDeletionTimestamp().IsZero() {
-		// if we are deleting the resources,
+	if !obj.GetDeletionTimestamp().IsZero() && r.DeletionCheck != nil {
```

Comment:
The initialization of this `DeletionCheck` is controlled by us, we just need to make sure it's getting init in declarative.WithCustomDeletionCheck, then we don't need to check this `nil` case, and will make the logic simplified.